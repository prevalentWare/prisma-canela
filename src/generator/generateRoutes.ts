import type { ParsedModel } from "../parser/types";
import { pascalCase } from "../utils/pascalCase";
import { camelCase } from "../utils/camelCase";
import type { Context } from "hono";
import type { ZodSchemaDetails } from "./types";

/**
 * Generates the content for a {modelName}.routes.ts file.
 * This file now only defines the OpenAPI routes and maps them to controller functions.
 */
export function generateRoutesFileContent(
  model: ParsedModel,
  zodSchemaInfo: ZodSchemaDetails
): string {
  const modelNamePascal = pascalCase(model.name);
  const modelNameCamel = camelCase(model.name);
  const modelNameLower = model.name.toLowerCase();
  const modelNamePluralLower = `${modelNameLower}s`;

  const idField = model.fields.find((f) => f.isId);
  // No longer need to return early, routes can be defined even if controller/service handles no-ID case.
  // The controller generation should handle not generating ID-based functions.

  // We still need ID info for route parameter definitions
  const idTypeIsNumber = idField?.type === "number";
  const idZodCoercionPipe = idTypeIsNumber
    ? `.pipe(z.coerce.number().int({ message: "Invalid ID format"}))`
    : "";
  const idZodType = idTypeIsNumber ? "number" : "string";

  // Updated imports: Remove service imports, add controller import
  const imports = [
    `import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi';`,
    `import { validator } from 'hono/validator';`,
    // `import type { Context } from 'hono';` // Context likely not needed directly here anymore
    ...zodSchemaInfo.imports, // Zod schema imports are still needed for createRoute
    `import * as controller from './controller';`, // Import controller functions
  ].join("\n");

  // Route definitions remain largely the same
  const routeDefinitions = `
// --- Route Definitions ---
// GET /${modelNamePluralLower}
const list${modelNamePascal}Route = createRoute({
  method: 'get',
  path: '/',
  tags: ['${modelNamePascal}'],
  summary: 'List all ${modelNamePluralLower}',
  responses: {
    200: {
      description: 'Returns a list of ${modelNamePluralLower}',
      content: {
        'application/json': {
          schema: z.array(${zodSchemaInfo.modelSchemaName}),
        },
      },
    },
    // TODO: Define 500 error response
  },
});

// Define ID-based routes only if idField exists
${
  idField
    ? `
// GET /${modelNamePluralLower}/{id}
const get${modelNamePascal}ByIdRoute = createRoute({
  method: 'get',
  path: '/{id}',
  tags: ['${modelNamePascal}'],
  summary: 'Get a ${modelNameLower} by ID',
  request: {
    params: z.object({
      id: z.string()${idZodCoercionPipe},
    }),
  },
  responses: {
    200: {
      description: 'Returns the ${modelNameLower}',
      content: {
        'application/json': {
          schema: ${zodSchemaInfo.modelSchemaName},
        },
      },
    },
    404: {
      description: '${modelNamePascal} not found',
      // TODO: Define standard error schema
    },
    // TODO: Define 400, 500 error responses
  },
});

// PATCH /${modelNamePluralLower}/{id}
const update${modelNamePascal}Route = createRoute({
  method: 'patch',
  path: '/{id}',
  tags: ['${modelNamePascal}'],
  summary: 'Update a ${modelNameLower} by ID',
  request: {
    params: z.object({
      id: z.string()${idZodCoercionPipe},
    }),
    body: {
      content: {
        'application/json': {
          schema: ${zodSchemaInfo.updateSchemaName},
        },
      },
    },
  },
  responses: {
    200: {
      description: 'Returns the updated ${modelNameLower}',
      content: {
        'application/json': {
          schema: ${zodSchemaInfo.modelSchemaName},
        },
      },
    },
    400: {
      description: 'Invalid input',
      // TODO: Define standard error schema
    },
    404: {
      description: '${modelNamePascal} not found',
      // TODO: Define standard error schema
    },
     // TODO: Define 500 error response
  },
});

// DELETE /${modelNamePluralLower}/{id}
const delete${modelNamePascal}Route = createRoute({
  method: 'delete',
  path: '/{id}',
  tags: ['${modelNamePascal}'],
  summary: 'Delete a ${modelNameLower} by ID',
  request: {
    params: z.object({
      id: z.string()${idZodCoercionPipe},
    }),
  },
  responses: {
    200: { // Consider 204 No Content if not returning the object
      description: '${modelNamePascal} deleted successfully',
      content: {
        'application/json': {
            // Optional: Return the deleted object's ID or a success message
            schema: z.object({ ${idField.name}: z.${idZodType}() }) // Use actual ID field name
        },
      },
    },
    404: {
      description: '${modelNamePascal} not found',
      // TODO: Define standard error schema
    },
    // TODO: Define 400, 500 error responses
  },
});
`
    : ""
} // End of conditional ID routes

// POST /${modelNamePluralLower} - Always defined
const create${modelNamePascal}Route = createRoute({
  method: 'post',
  path: '/',
  tags: ['${modelNamePascal}'],
  summary: 'Create a new ${modelNameLower}',
  request: {
    body: {
      content: {
        'application/json': {
          schema: ${zodSchemaInfo.createSchemaName},
        },
      },
    },
  },
  responses: {
    201: {
      description: 'Returns the created ${modelNameLower}',
      content: {
        'application/json': {
          schema: ${zodSchemaInfo.modelSchemaName},
        },
      },
    },
    400: {
      description: 'Invalid input',
      // TODO: Define standard error schema
    },
    // TODO: Define 500 error response
  },
});
`;

  // Updated app setup: Map routes to controller functions using standard Hono methods
  const appSetup = `
// --- Hono App Setup ---
// Using standard Hono methods for clearer middleware application
const ${modelNameCamel}Routes = new OpenAPIHono(); // Keep OpenAPIHono for potential doc gen later

// GET /${modelNamePluralLower} (List)
${modelNameCamel}Routes.get(
  '/',
  controller.list${modelNamePascal}
);

// POST /${modelNamePluralLower} (Create)
${modelNameCamel}Routes.post(
  '/',
  validator('json', (value, c) => {
    // Reuse the schema defined in createRoute for validation
    const parsed = create${modelNamePascal}Route.request.body.content['application/json'].schema.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data; // Make validated data available via c.req.valid('json')
  }),
  controller.create${modelNamePascal}
);

${
  idField
    ? `
// GET /${modelNamePluralLower}/{id}
${modelNameCamel}Routes.get(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = get${modelNamePascal}ByIdRoute.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.get${modelNamePascal}ById
);

// PATCH /${modelNamePluralLower}/{id}
${modelNameCamel}Routes.patch(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = update${modelNamePascal}Route.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  validator('json', (value, c) => {
    const parsed = update${modelNamePascal}Route.request.body.content['application/json'].schema.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.update${modelNamePascal}
);

// DELETE /${modelNamePluralLower}/{id}
${modelNameCamel}Routes.delete(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = delete${modelNamePascal}Route.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.delete${modelNamePascal}
);
`
    : ""
} // Conditionally register ID-based routes

// TODO: Optionally re-attach OpenAPI route definitions for documentation generation if needed
// Example: ${modelNameCamel}Routes.doc('/path', list${modelNamePascal}Route.getOpenAPIMetadata());

export default ${modelNameCamel}Routes;
`;

  // Return only imports, route definitions, and app setup
  return [
    imports,
    "\n",
    routeDefinitions,
    // handlers variable removed
    "\n",
    appSetup,
  ].join("\n");
}
