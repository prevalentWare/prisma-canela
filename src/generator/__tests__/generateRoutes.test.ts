import { describe, it, expect } from "vitest";
import type { ParsedModel } from "../../parser/types"; // Adjusted import path
import { generateRoutesFileContent } from "../generateRoutes"; // Adjusted import path
import type { ZodSchemaDetails } from "../types"; // Import from shared types

// Mock Data
const mockUserModel: ParsedModel = {
  name: "User",
  dbName: "users",
  fields: [
    {
      name: "id",
      type: "number",
      kind: "scalar",
      isList: false,
      isRequired: true,
      isUnique: false,
      isId: true,
      hasDefaultValue: true,
    },
    {
      name: "email",
      type: "string",
      kind: "scalar",
      isList: false,
      isRequired: true,
      isUnique: true,
      isId: false,
      hasDefaultValue: false,
    },
    {
      name: "name",
      type: "string",
      kind: "scalar",
      isList: false,
      isRequired: false,
      isUnique: false,
      isId: false,
      hasDefaultValue: false,
    },
    // Example enum field
    {
      name: "role",
      type: "enum",
      kind: "enum",
      enumName: "Role", // Assume an enum Role exists
      isList: false,
      isRequired: true,
      isUnique: false,
      isId: false,
      hasDefaultValue: false,
    },
  ],
};

const mockZodSchemaInfo: ZodSchemaDetails = {
  imports: [
    `import { UserSchema, createUserSchema, updateUserSchema } from './schema';`,
    `import { Role } from '@prisma/client';`, // Assuming enum comes from prisma client
  ],
  modelSchemaName: "UserSchema",
  createSchemaName: "createUserSchema",
  updateSchemaName: "updateUserSchema",
};

describe("generateRoutesFileContent", () => {
  it("should generate correct refactored route file content for a User model", () => {
    const result = generateRoutesFileContent(mockUserModel, mockZodSchemaInfo);

    // Check for controller import
    expect(result).toContain("import * as controller from './controller';");
    // Check Zod/Hono imports still exist
    expect(result).toContain(
      "import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi';"
    );
    expect(result).toContain("from './schema';");
    // Check service import is REMOVED
    expect(result).not.toContain("from './service';");
    // Check handler definitions are REMOVED
    expect(result).not.toContain("const handleListUser = async");
    expect(result).not.toContain("const handleGetUserById = async");
    // Check route definitions still exist
    expect(result).toContain("const listUserRoute = createRoute");
    expect(result).toContain("const getUserByIdRoute = createRoute");
    expect(result).toContain("const createUserRoute = createRoute");
    expect(result).toContain("const updateUserRoute = createRoute");
    expect(result).toContain("const deleteUserRoute = createRoute");
    // Check routes are mapped to controller functions using standard Hono methods and validator
    expect(result).toMatch(/userRoutes\.get\(\s*'\/',\s*controller\.listUser/);
    expect(result).toMatch(
      /userRoutes\.post\(\s*'\/',\s*validator\('json'/ // Check POST starts correctly
    );
    expect(result).toMatch(/\s*controller\.createUser/); // Check controller follows validator
    expect(result).toMatch(
      /userRoutes\.get\(\s*'\/{id}',\s*validator\('param'/ // Check GET by ID starts correctly
    );
    expect(result).toMatch(/\s*controller\.getUserById/); // Check controller follows validator
    expect(result).toMatch(
      /userRoutes\.patch\(\s*'\/{id}',\s*validator\('param'/ // Check PATCH starts correctly
    );
    expect(result).toMatch(/\s*validator\('json'/); // Check PATCH includes json validator
    expect(result).toMatch(/\s*controller\.updateUser/); // Check controller follows validators
    expect(result).toMatch(
      /userRoutes\.delete\(\s*'\/{id}',\s*validator\('param'/ // Check DELETE starts correctly
    );
    expect(result).toMatch(/\s*controller\.deleteUser/); // Check controller follows validator

    expect(result).toMatchSnapshot();
  });

  it("should generate only non-ID routes if model has no ID field", () => {
    const modelWithoutId: ParsedModel = {
      name: "LogEntry",
      dbName: null,
      fields: [
        {
          name: "message",
          type: "string",
          kind: "scalar",
          isList: false,
          isRequired: true,
          isUnique: false,
          isId: false,
          hasDefaultValue: false,
        },
        {
          name: "timestamp",
          type: "date",
          kind: "scalar",
          isList: false,
          isRequired: true,
          isUnique: false,
          isId: false,
          hasDefaultValue: true,
        },
      ],
    };
    // Provide necessary Zod info even if updateSchema isn't used in routes
    const logEntryZodInfo: ZodSchemaDetails = {
      imports: [
        `import { LogEntrySchema, createLogEntrySchema, updateLogEntrySchema } from './schema';`,
      ],
      modelSchemaName: "LogEntrySchema",
      createSchemaName: "createLogEntrySchema",
      updateSchemaName: "updateLogEntrySchema",
    };

    const result = generateRoutesFileContent(modelWithoutId, logEntryZodInfo);

    // Check controller import is present
    expect(result).toContain("import * as controller from './controller';");
    // Check LIST and CREATE routes ARE defined
    expect(result).toContain("const listLogEntryRoute = createRoute");
    expect(result).toContain("const createLogEntryRoute = createRoute");
    // Check LIST and CREATE routes ARE mapped using standard Hono methods
    expect(result).toMatch(
      /logEntryRoutes\.get\(\s*'\/',\s*controller\.listLogEntry/ // Check GET list
    );
    expect(result).toMatch(
      /logEntryRoutes\.post\(\s*'\/',\s*validator\('json'/ // Check POST create with validator
    );
    expect(result).toMatch(/\s*controller\.createLogEntry/); // Check controller follows validator

    // Check ID-based routes are NOT defined
    expect(result).not.toContain("const getLogEntryByIdRoute = createRoute");
    expect(result).not.toContain("const updateLogEntryRoute = createRoute");
    expect(result).not.toContain("const deleteLogEntryRoute = createRoute");
    // Check ID-based routes are NOT mapped
    expect(result).not.toContain("openapi(getLogEntryByIdRoute");
    expect(result).not.toContain("openapi(updateLogEntryRoute");
    expect(result).not.toContain("openapi(deleteLogEntryRoute");

    expect(result).toMatchSnapshot();
  });
});
