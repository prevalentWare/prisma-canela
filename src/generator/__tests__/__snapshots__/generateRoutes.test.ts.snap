// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRoutesFileContent > should generate correct refactored route file content for a User model 1`] = `
"import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi';
import { validator } from 'hono/validator';
import { UserSchema, createUserSchema, updateUserSchema } from './schema';
import { Role } from '@prisma/client';
import * as controller from './controller';



// --- Route Definitions ---
// GET /users
const listUserRoute = createRoute({
  method: 'get',
  path: '/',
  tags: ['User'],
  summary: 'List all users',
  responses: {
    200: {
      description: 'Returns a list of users',
      content: {
        'application/json': {
          schema: z.array(UserSchema),
        },
      },
    },
    // TODO: Define 500 error response
  },
});

// Define ID-based routes only if idField exists

// GET /users/{id}
const getUserByIdRoute = createRoute({
  method: 'get',
  path: '/{id}',
  tags: ['User'],
  summary: 'Get a user by ID',
  request: {
    params: z.object({
      id: z.string().pipe(z.coerce.number().int({ message: "Invalid ID format"})),
    }),
  },
  responses: {
    200: {
      description: 'Returns the user',
      content: {
        'application/json': {
          schema: UserSchema,
        },
      },
    },
    404: {
      description: 'User not found',
      // TODO: Define standard error schema
    },
    // TODO: Define 400, 500 error responses
  },
});

// PATCH /users/{id}
const updateUserRoute = createRoute({
  method: 'patch',
  path: '/{id}',
  tags: ['User'],
  summary: 'Update a user by ID',
  request: {
    params: z.object({
      id: z.string().pipe(z.coerce.number().int({ message: "Invalid ID format"})),
    }),
    body: {
      content: {
        'application/json': {
          schema: updateUserSchema,
        },
      },
    },
  },
  responses: {
    200: {
      description: 'Returns the updated user',
      content: {
        'application/json': {
          schema: UserSchema,
        },
      },
    },
    400: {
      description: 'Invalid input',
      // TODO: Define standard error schema
    },
    404: {
      description: 'User not found',
      // TODO: Define standard error schema
    },
     // TODO: Define 500 error response
  },
});

// DELETE /users/{id}
const deleteUserRoute = createRoute({
  method: 'delete',
  path: '/{id}',
  tags: ['User'],
  summary: 'Delete a user by ID',
  request: {
    params: z.object({
      id: z.string().pipe(z.coerce.number().int({ message: "Invalid ID format"})),
    }),
  },
  responses: {
    200: { // Consider 204 No Content if not returning the object
      description: 'User deleted successfully',
      content: {
        'application/json': {
            // Optional: Return the deleted object's ID or a success message
            schema: z.object({ id: z.number() }) // Use actual ID field name
        },
      },
    },
    404: {
      description: 'User not found',
      // TODO: Define standard error schema
    },
    // TODO: Define 400, 500 error responses
  },
});
 // End of conditional ID routes

// POST /users - Always defined
const createUserRoute = createRoute({
  method: 'post',
  path: '/',
  tags: ['User'],
  summary: 'Create a new user',
  request: {
    body: {
      content: {
        'application/json': {
          schema: createUserSchema,
        },
      },
    },
  },
  responses: {
    201: {
      description: 'Returns the created user',
      content: {
        'application/json': {
          schema: UserSchema,
        },
      },
    },
    400: {
      description: 'Invalid input',
      // TODO: Define standard error schema
    },
    // TODO: Define 500 error response
  },
});




// --- Hono App Setup ---
// Using standard Hono methods for clearer middleware application
const userRoutes = new OpenAPIHono(); // Keep OpenAPIHono for potential doc gen later

// GET /users (List)
userRoutes.get(
  '/',
  controller.listUser
);

// POST /users (Create)
userRoutes.post(
  '/',
  validator('json', (value, c) => {
    // Reuse the schema defined in createRoute for validation
    const parsed = createUserRoute.request.body.content['application/json'].schema.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data; // Make validated data available via c.req.valid('json')
  }),
  controller.createUser
);


// GET /users/{id}
userRoutes.get(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = getUserByIdRoute.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.getUserById
);

// PATCH /users/{id}
userRoutes.patch(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = updateUserRoute.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  validator('json', (value, c) => {
    const parsed = updateUserRoute.request.body.content['application/json'].schema.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.updateUser
);

// DELETE /users/{id}
userRoutes.delete(
  '/{id}',
  validator('param', (value, c) => {
    const parsed = deleteUserRoute.request.params.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data;
  }),
  controller.deleteUser
);
 // Conditionally register ID-based routes

// TODO: Optionally re-attach OpenAPI route definitions for documentation generation if needed
// Example: userRoutes.doc('/path', listUserRoute.getOpenAPIMetadata());

export default userRoutes;
"
`;

exports[`generateRoutesFileContent > should generate only non-ID routes if model has no ID field 1`] = `
"import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi';
import { validator } from 'hono/validator';
import { LogEntrySchema, createLogEntrySchema, updateLogEntrySchema } from './schema';
import * as controller from './controller';



// --- Route Definitions ---
// GET /logentrys
const listLogEntryRoute = createRoute({
  method: 'get',
  path: '/',
  tags: ['LogEntry'],
  summary: 'List all logentrys',
  responses: {
    200: {
      description: 'Returns a list of logentrys',
      content: {
        'application/json': {
          schema: z.array(LogEntrySchema),
        },
      },
    },
    // TODO: Define 500 error response
  },
});

// Define ID-based routes only if idField exists
 // End of conditional ID routes

// POST /logentrys - Always defined
const createLogEntryRoute = createRoute({
  method: 'post',
  path: '/',
  tags: ['LogEntry'],
  summary: 'Create a new logentry',
  request: {
    body: {
      content: {
        'application/json': {
          schema: createLogEntrySchema,
        },
      },
    },
  },
  responses: {
    201: {
      description: 'Returns the created logentry',
      content: {
        'application/json': {
          schema: LogEntrySchema,
        },
      },
    },
    400: {
      description: 'Invalid input',
      // TODO: Define standard error schema
    },
    // TODO: Define 500 error response
  },
});




// --- Hono App Setup ---
// Using standard Hono methods for clearer middleware application
const logEntryRoutes = new OpenAPIHono(); // Keep OpenAPIHono for potential doc gen later

// GET /logentrys (List)
logEntryRoutes.get(
  '/',
  controller.listLogEntry
);

// POST /logentrys (Create)
logEntryRoutes.post(
  '/',
  validator('json', (value, c) => {
    // Reuse the schema defined in createRoute for validation
    const parsed = createLogEntryRoute.request.body.content['application/json'].schema.safeParse(value);
    if (!parsed.success) {
      return c.json({ error: 'Validation failed', issues: parsed.error.issues }, 400);
    }
    return parsed.data; // Make validated data available via c.req.valid('json')
  }),
  controller.createLogEntry
);

 // Conditionally register ID-based routes

// TODO: Optionally re-attach OpenAPI route definitions for documentation generation if needed
// Example: logEntryRoutes.doc('/path', listLogEntryRoute.getOpenAPIMetadata());

export default logEntryRoutes;
"
`;
